### DISCUSSION TITLE
Design system performance tuning

### DISCUSSION THREAD

**Discussion by loliver** [vanilla-extract-css/vanilla-extract#156] [Q&A] [answered]
Hi all,

We're having a discussion internally on how to handle splitting our themes to produce smaller css files. Presently with the default usage of `createTheme` on our system (6 themes, presently 25 or so components but growing) we are producing a 593kb css file.

One option we're looking at is using `createTheme` lower down our tokens tree, so instead of `createTheme(themeObject)` we'd be using `createTheme(themeObject.components.componentName)` and having separate themes created per component. This then comes with the issue of having to bootstrap consuming apps with every component they use to get theme classes applied.

Has anyone looked in to doing this sort of performance tuning on their design system? I know I read somewhere that the authors had thought about trying to attempt tree shaking the generated vars and/or classes with the Webpack plugin but it was very far in the too hard basket for now.

Cheers,
Oli

**Comment by mattcompiles** [comment, marked as answer]
Hey @loliver 👋 

Not sure where you heard about tree shaking not working. That issue was specific to treat rather than vanilla-extract. You should be able to tree shake your themes as long as you're using the [createThemeContract](https://vanilla-extract.style/documentation/styling-api/#createthemecontract) API. If you've done that, make sure each theme lives in its own file. Then normal CSS splitting techniques should apply.

**Reply by loliver** [reply]
Ahh yeah that was probably where I saw it, cheers for the heads up, will have a play when I get the chance.

**Reply by loliver** [reply]
How do I set up Webpack cacheGroups to match the generated css files relative to where my css.ts files (which are getting matched and split as expected) live?

I have my themeContract being generated in a separate file, which is then imported to the 4 theme files. The tokens used to generate every theme are also in separate files.

**Reply by mattcompiles** [reply]
Honestly not 100% sure about this one mate. I'd be interested in your findings though. However, each `.css.ts` file that produces CSS (`createThemeContract` doesn't produce any) should create a matching virtual CSS file. e.g. `styles.css.ts` -> `styles.vanilla.css`. In webpack we use a feature called [matchResource](https://webpack.js.org/api/loaders/#inline-matchresource) to do this.

**Reply by loliver** [reply]
Thanks for the link, will give it a read and see if I can gleam any nuggets of wisdom from it.

**Reply by loliver** [reply]
I added the [vendor chunk split from here](https://medium.com/hackernoon/the-100-correct-way-to-split-your-chunks-with-webpack-f8a9df5b7758) and it looks like Webpack sees the generated css as living under Vanilla Extract rather than the package we have our design tokens in, going to figure out how to rename our Style Dictionary exports and see if I can match purely by filename and see if that helps.

**Reply by loliver** [reply]
Got it working!

The following cache group splits all of the generated sheets based on `debugId`, would be cool if there was some way to put a more human readable identifier on there to split by when you generate the css with the plugin.

```js
      vendor: {
        test: /[\\/]node_modules[\\/]/,
        name(module) {
          // get the name. E.g. node_modules/packageName/not/this/part.js
          // or node_modules/packageName
          const packageMatch = module.context.match(
            /[\\/]node_modules[\\/](.*?)([\\/]|$)/
          )
          const packageName =
            packageMatch.indexOf('@vanilla-extract') > -1
              ? packageMatch[1] + '.' + module.debugId
              : packageMatch[1]

          // npm package names are URL-safe, but some servers don't like @ symbols
          return `npm.${packageName.replace('@', '')}`
        },
        priority: 0
      }
```

That's given me the following files:

- npm.vanilla-extract.2920.css
- npm.vanilla-extract.2919.css
- npm.vanilla-extract.2612.css
- npm.vanilla-extract.2917.css
- npm.vanilla-extract.2921.css

**Reply by mattcompiles** [reply]
Nice work. Thanks for reporting your findings. I agree this definitely should be easier though. Will have a look into it.

**Reply by mattcompiles** [reply]
Had a look into this. Looks like best way to target CSS extracted via vanilla for code splitting is using the following.

```
cacheGroups: {
  myCss: {
    test: (m) => {
      if (m.type === 'css/mini-extract') {
		const cssResource = m.issuer.matchResource; // -> 'MyApp/src/style.css.ts.vanilla.css'
        // Replace with whatever check you want
        return /some-test/.test(cssResource)
      }
    },
  }
}
```
